### DISCUSSION TITLE
source files for Ctrlr

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#259] [Q&A]
Where can I find the source code files for Ctrlr v 5.3.201? I'm asking this because I managed to compile for Mac the latest "master" build (using Juce 6.0.8) published by Roman Kubiak about a week ago. This version seems fit for further investigations, because now on my Mac MIDI in/out is at least working, which was not the case with v5.6.0 (using Juce 6.0.7). But I also want to try to compile v5.3.201 for the Mac, which does not exist, at least to my knowledge, on the Crtrlr.org site, but seems quite a favourite among Windows users. So can anyone help me out? Any questions on compiling for Mac I'm willing to answer, but mind you I'm still learning! (As a little joke aside: it is now possible to make a Dutch version of Ctrlr...)

**Comment by Godlike-Productions** [comment]
You can find them here - https://github.com/RomanKubiak/ctrlr/tree/73699fddc141f0e71c26adac4392860b06bc40d2

This was the state of the Repo at the date 5.3.201 was uploaded to nightly.

**Reply by dobo365** [reply]
I downloaded the zip of that commit. Now will load it into Visual Studio and try to find how to rebuild this.
Thx a lot for pointing to this particular version that I was looking for.

**Reply by shooking** [reply]
How do you find the tag/release/timestamp for a specific release in the git please?
Normally releases are made in the git itself.
When I look I can see following branches:
master (default) - so git clone will pull that state right?
kiosk
rpi_tests
remove_cmake (stale branch)

Within the master I can see following releases:
5.6.0
5.5.9
5.5.1
but not this 5.3.1 - so how did you work out the timestamp please?
Curious because I would love to submit a pull request for building on Pi - but to do this it makes sense to start at a known point, test it, then apply for push back into that branch/release.

**Comment by dobo365** [comment]
On my Mac VM I'm using 5.3.198 without any issue. Tbh, I don't know the differences with 5.3.201 that I'm using as main dev tool on my PC.
What I would like to have now it is a 5.3.201 (or 198) Mac version for M1 processors and running on Big Sur.

**Reply by Godlike-Productions** [reply]
We are recompiling it so we can apply iLok copy protection to a plugin. I need to add some  linker flags to allow enough space for the wrapper to add code. If I end up with a M1 mac, I'll try to recompile on that.  OSX is proving to be a nightmare to support.

**Reply by damiensellier** [reply]
This iLok protection addon would be a huge improvement. I was thinking about copy protection with user/serial for further commercial projects. This is a great idea. Since ctrlr is based on Juce libs and Pace iLok acquired JUCE, the implementation should be flawless. I'll check this thread if ever you found out how to recompile everything with iLok capabilities and this damn mac m1 support. If it's not posssible I'll go with JUCE only for heavier commercial plugins.

**Reply by Godlike-Productions** [reply]
With windows, the iLok wrapping is easy and flawless.  On OSX it must have the linker flags otherwise it fails on the AU.  The standalone does wrap ok though.

I haven't attempted the deeper code hooks methods though.  It's beyond my current capabilities and would require patching of the ctrlr code, prior to compilation, and may require it to be compiled seperately for each protected product.

Wrapping is the go for now.

**Comment by dobo365** [comment]
Will try recompiling it from my side.
Did you rebuild it with Visual Studio? Version 2010 or last one?
Could be good to also add some versioning to not confuse us. Like 5.3.201.1
and so on
Great!

**Reply by Godlike-Productions** [reply]
If you manage to recompile can you please, please, please add these extra Linker Flags. This allows for iLok copy protection.

It won't impact normal use of the program, it just creates some extra space in the program header.

Extra Linker Flags -Xlinker -headerpad -Xlinker 800

**Reply by bijlevel** [reply]
The compiling was attempted for Mac using Xcode. The standalone build succeeded, but the AU plugin did not. And that was the main reason for (re?)compiling this old version. I will now focus on the latest version (5.6.23). The compiling for Mac did succeed, including the AU plugin. I will now try to compile this version for Windows using VS 2019, unless you did start with this "job" ;-).

**Comment by bijlevel** [comment]
As I wrote in my opening post, I did manage to compile the latest master (5.6.23) for Mac, including the AU with the extra linker flags. BUT this version gives "no matching overloads found" as soon as a Lua function tries to send a Ctrlr MIDI sysex message. At least that's what I could test. Sending a sysex message straight from the panel editor does work, though. If this error could be resolved, then MAYBE we could have a working Mac version. But I'm afraid other problems will creep up...

I did download the zip of the commit for v5.3.198 (or 201?) (in the source code it is named v5.4.0...), but until now trying to compile this one is for sure a nightmare!! (I almost never use 2 acclamations but this time I do ;-)). So many compiler issues come up, many completely adacadabra to me. Also before I can even begin to compile I have to reorganise many files and add JUCE files in the various directories, so at least the compiler can find them. Also this zip doesn't contain AFAIK the AU wrapper mentioned by Roman Kubiak in the README.md file, so probably also no chance for a AU component.

**Reply by Godlike-Productions** [reply]
Bugger

**Reply by Ted-juh** [reply]
Yeah, been there, done that. But for the windows version. Don't forget to extract boost to the right folder. And I think you have to compile the introjucer, before compiling Ctrlr as well.

**Reply by bijlevel** [reply]
Yep, I know about unzipping 'boost'. It is the first error that comes up when compiling. I don't know what you mean by compiling Introjucer first. The latest JUCE download for Mac comes with Projucer pre-built. I did try to add the Ctrlr files to a new Projucer project, but when adding the boost directory the Projucer hangs, not responding anymore. So I did compare the files used by Projucer for the latest Ctrlr build (v5.6.23, added about two weeks ago and comes with Ctrlr.jucer), and I discovered it does not contain the boost directory. So that one is probably not needed in the Projucer project, but apparently should be present in the project folder. Well, I will keep on trying, but on the other hand would rather see the errors fixed in the latest master, because that one does compile OK and builds also the AU plugin.

**Reply by Ted-juh** [reply]
To my knowledge:
 V5.3.201 for windows is from **march (29?) 2016.** It used **Juce V4.2.0** with "IntroJucer".  
V5.3.198 for Mac should have been released around that time as well if you ask me. 

Your version dates from **March 2017**, which uses **Juce V4.30**. IntroJucer was dropped by then, only projucer was available. 
So the question is, do you really have the right source files for version 5.3.198?

**Reply by bijlevel** [reply]
I'm not sure if you're right with the dates. This is what I see on the ctrlr.org site:

<img width="728" alt="Screenshot 2021-04-25 at 20 09 04" src="https://user-images.githubusercontent.com/11859363/116004331-8f583480-a602-11eb-95fb-ca6daec66d92.png">

So they both date from April 7, 2017 and the commit is from March 30, 2017. You're right about Projucer.

**Reply by Ted-juh** [reply]
https://web.archive.org/web/20160503065043/http://ctrlr.org/nightly/?C=M;O=D

March 15 2016

*Edited to add the date.

**Reply by bijlevel** [reply]
Hmm, quite confusing, but you're right! At further inspection of the dates, many versions have the same date on ctrlr.org! I should have looked better. I will go back further in time on the commits to find the real version 5.3.198 for Mac! Thanks for pointing this out.

**Reply by Ted-juh** [reply]
Well, there is just one commit on march 15 2016 and it explicitly states "Mac warning fixes". 

Good luck.

**Reply by bijlevel** [reply]
I did find the source files and succeed to compile v5.3.198 for Mac, but only Ctrlr standalone. Compiling the AU plugin posed too many problems, because it relies on the CoreAudio SDK for Mac. For various reasons installing and using this SDK on my Mac was a real PITA and I gave up. For now I'm focusing on the latest version published by Roman Kubiak. It did compile fine on my Mac, including the AU plugin and after solving the initial errors it does run panels made with older versions of Ctrlr. I will now focus on compiling it for Windows 10 using VS 2019, unless somebody else did this already.

**Reply by Ted-juh** [reply]
I managed to compile version 5.3.201 on windows 10 with VS 2010. Didn't succeed compiling the latest version. But keep in mind that there was a problem with midi on Roman's version. Somebody made a xommit where the midi was working if I remember correctly.

If you succeed to compile for Windows 10, would you be so kind to share how you did it? Lots of people are dying to know..

**Reply by bijlevel** [reply]
I did succeed to compile Ctrlr Standalone v5.6.23 (the latest master) for Windows 10 x64 with Visual Studio 2019. I did a very short test, but at least Ctrlr does connect MIDI in/out with my Midi interface. I will share how I did it, but need time to write down the steps I took in a way that even I understand it myself ;-)! And also I will do some more testing before calling it a succes...
